### PR TITLE
Allow streaming interface to CBOR

### PR DESCRIPTION
**Pull request by mildred**
See #2 for rationale.

### CODE REVIEW

**Comment by mildred** [comment]
I am also working on go-ipld. These modifications are necessary for what I'm doing there (see ipfs/go-ipld#17). I haven't pushed any commit there yet but basically I need to be able to handle the CBOR structure as it is read from the underlying `io.Reader`. I don't want to wait until the complete CBOR document is read, and I don't want to store the complete data structure in memory because the streaming interface already allows me to work on the data bit by bit.

I believe these commit won't change much now.

@whyrusleeping what do you think?

**Comment by jbenet** [comment]
@whyrusleeping could you please take a look here? or give me collab access?

File: go/decoder.go [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+}
+
+func (dec *Decoder) DecodeAny(v DecodeValue) error {
+	var didread int
+	var err error
+
+	_, err = io.ReadFull(dec.rin, dec.c)
+
+	if didread == 1 {
+		/* log.Printf("got one %x\n", dec.c[0]) */
```
Comment (whyrusleeping) [inline go/decoder.go]: what is this case for?

File: go/decoder.go [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+	}
+
+	if err != nil {
```
Comment (whyrusleeping) [inline go/decoder.go]: should move this closer to where the error is created

File: go/decoder.go [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+	rv := reflect.ValueOf(v)
+
+	return dec.DecodeAny(newReflectValue(rv))
+}
+
+func (dec *Decoder) DecodeAny(v DecodeValue) error {
+	var didread int
+	var err error
+
+	_, err = io.ReadFull(dec.rin, dec.c)
```
Comment (whyrusleeping) [inline go/decoder.go]: just use type inference here:

```
_, err := io.ReadFull(...
```

File: go/decoder.go [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+		return aux, nil
+	} else if cborInfo == int8Follows {
+		didread, err := io.ReadFull(dec.rin, dec.b8[:1])
+		if didread == 1 {
+			aux = uint64(dec.b8[0])
+		}
+		return aux, err
+	} else if cborInfo == int16Follows {
+		didread, err := io.ReadFull(dec.rin, dec.b8[:2])
+		if didread == 2 {
```
Comment (whyrusleeping) [inline go/decoder.go]: this sort of error handling seems weird to me... why arent we checking the error? `ReadFull` is guaranteed to read as much as we ask it to, and will error in any other case.

File: go/decoder.go [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+func (dec *Decoder) innerDecodeC(rv DecodeValue, c byte) error {
+	cborType := c & typeMask
+	cborInfo := c & infoBits
+
+	aux, err := dec.handleInfoBits(cborInfo)
+	if err != nil {
+		log.Printf("error in handleInfoBits: %v", err)
+		return err
+	}
+	//log.Printf("cborType %x cborInfo %d aux %x", cborType, cborInfo, aux)
```
Comment (whyrusleeping) [inline go/decoder.go]: remove commented out code

File: go/decoder.go [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+	return dec.innerDecodeC(v, dec.c[0])
+}
+
+func (dec *Decoder) handleInfoBits(cborInfo byte) (uint64, error) {
+	var aux uint64
+
+	if cborInfo <= 23 {
+		aux = uint64(cborInfo)
+		return aux, nil
+	} else if cborInfo == int8Follows {
```
Comment (whyrusleeping) [inline go/decoder.go]: make this into a switch statement?

File: go/decoder.go [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+	cborInfo := c & infoBits
+
+	aux, err := dec.handleInfoBits(cborInfo)
+	if err != nil {
+		log.Printf("error in handleInfoBits: %v", err)
+		return err
+	}
+	//log.Printf("cborType %x cborInfo %d aux %x", cborType, cborInfo, aux)
+
+	if cborType == cborUint {
```
Comment (whyrusleeping) [inline go/decoder.go]: switch statements

File: go/decoder.go [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+		} else if cborInfo == int64Follows {
+			d := math.Float64frombits(aux)
+			return rv.SetFloat64(d)
+		} else if cborInfo == cborFalse {
+			return rv.SetBool(false)
+		} else if cborInfo == cborTrue {
+			return rv.SetBool(true)
+		} else if cborInfo == cborNull {
+			return rv.SetNil()
+		}
```
Comment (whyrusleeping) [inline go/decoder.go]: when switching to a switch statement, make sure to have a default case to catch errors

**Comment by whyrusleeping** [comment]
lots of comments, this is a big changeset that i'm not convinced we need. What is the usecase? Are cbor documents going to be very large that holding them in memory is going to be an issue?

Also, with a changeset like this, adding tests is going to be important.

**Comment by mildred** [comment]
If you look at the changes, most of it is code that has been moved. Everything you commented on was code that was already there in another file. That's why I tried to split into multiple commits.

The use case is to have a streaming interface where we process events instead. It can be used to avoid keeping more data than needed in memory, have the ability to process the data even if it hasn't been all received yet.

The use case is also to be able to have detailed control over the generated data structure.

In practice, this is used by ipfs/go-ipld#17 that implements a streaming interface to go-ipld. Even without the streaming interface, it could be used to customize the data structure. Generate `ipld.Node` instead of `map[interface{}]interface{}`.

**Comment by whyrusleeping** [comment]
Could you not split it up into separate files then? It makes it really difficult to review this when it looks like everything was removed and rewritten

**Comment by jbenet** [comment]
@whyrusleeping maybe try looking at the last two commits themselves?

(FWIW, i think separating to many files is a very good idea. But, this code comes from elsewhere (maybe the readme should say so) and separating will make it much harder to merge any upstream changes. unless we want to forsake that option, which may be valid, i dont recall it being maintained. --- @whyrusleeping where was it from?)

**Comment by whyrusleeping** [comment]
I'm fine with splitting it into separate files, but for the purposes of review, it makes things difficult. I  guess I can just look at the last two commits

**Comment by whyrusleeping** [comment]
Alright, If i can get a couple of tests to at least test the new behaviour (and maybe one to make sure the old behaviour still works)  i'll merge this so i'm no longer blocking ipld stuff.

**Comment by mildred** [comment]
The last commit should produce a diff that is a little easier to review. That's quite unfortunate that diff tools can't show that in any meaningful way.

**Comment by whyrusleeping** [comment]
oh wow, that does look much better. good stuff!

**Pull request merged**